crypto/objects/obj_xref.c

```diff
@@ -40,69 +42,108 @@ static int sigx_cmp(const nid_triple *const *a, const nid_triple *const *b)
 
 IMPLEMENT_OBJ_BSEARCH_CMP_FN(const nid_triple *, const nid_triple *, sigx);
 
+static CRYPTO_ONCE sig_init = CRYPTO_ONCE_STATIC_INIT;
+
+DEFINE_RUN_ONCE_STATIC(o_sig_init)
+{
+    sig_app = sk_nid_triple_new(sig_sk_cmp);
+    sigx_app = sk_nid_triple_new(sigx_cmp);
+    sig_lock = CRYPTO_THREAD_lock_new();
+
+    if (sig_app == NULL || sigx_app == NULL || sig_lock == NULL) {
+        sk_nid_triple_free(sig_app);
+        sk_nid_triple_free(sigx_app);
+        CRYPTO_THREAD_lock_free(sig_lock);
+        sig_app = NULL;
+        sigx_app = NULL;
+        sig_lock = NULL;
+        return 0;
+    }
+    return 1;
+}
+
+static int obj_sig_init(void)
+{
+    return RUN_ONCE(&sig_init, o_sig_init);
+}
+
 int OBJ_find_sigid_algs(int signid, int *pdig_nid, int *ppkey_nid)
 {
     nid_triple tmp;
     const nid_triple *rv = NULL;
+    int idx, lock = sig_app != NULL;
+
     tmp.sign_id = signid;
 
-    if (sig_app != NULL) {
-        int idx = sk_nid_triple_find(sig_app, &tmp);
-        rv = sk_nid_triple_value(sig_app, idx);
+    if (lock) {
+        if (!CRYPTO_THREAD_read_lock(sig_lock))
+            return 0;
+        idx = sk_nid_triple_find(sig_app, &tmp);
+        if (idx >= 0)
+            rv = sk_nid_triple_value(sig_app, idx);
     }
 #ifndef OBJ_XREF_TEST2
```

Comment:
What is this??

Reply:
Dunno, existing mess I dared not touch.  I expect it is always included since the code doesn't make sense otherwise.

Reply:
I think it can just be removed.

Reply:
I've removed it.